docs/trace/building-your-own-processor/Program.cs

```diff
@@ -25,20 +26,46 @@ public class Program
 
     public static void Main()
     {
-        using var otel = Sdk.CreateTracerProvider(b => b
-            .AddActivitySource("MyCompany.MyProduct.MyLibrary")
+        /* a simple case, where default ALWAYS_ON sampler is used:
+        using var tracerProvider = Sdk.CreateTracerProvider()
+            .AddProcessor(new BatchingActivityProcessor(new ZipkinExporter()));
+        */
 
-            // TODO: seems buggy as ShutdownAsync is called 6 times
-            // TODO: need to discuss the expectation, currently FlushAsync is not called by default
-            // TODO: should the dispose order be C, B, A or A, B C?
-            .AddProcessorPipeline(p => p.AddProcessor(current => new MyActivityProcessor("A")))
-            .AddProcessorPipeline(p => p.AddProcessor(current => new MyActivityProcessor("B")))
-            .AddProcessorPipeline(p => p.AddProcessor(current => new MyActivityProcessor("C"))));
+        /* making it easier for the user:
+        using var tracerProvider = Sdk.CreateTracerProvider()
+            .AddZipkinExporter(options);
 
-        using (var activity = MyActivitySource.StartActivity("SayHello"))
+        1) AddExporter should be an extension function provided by individual exporter
+        2) individual exporter should pick the right (batching vs. simple) processor by default
+        3) individual exporter can expose an option letting people to use a non-default processor
+        */
+
+        using var tracerProvider = Sdk.CreateTracerProvider(new ProbabilitySampler(0.5))
+            .AddListener("MyCompany.MyProduct.MyLibrary")
+            .AddListener("MyCompany.AnotherProduct.*")
+            .AddProcessor(new MyActivityProcessor("A"))

```

Comment:
I was looking at this and how we built `ActivityProcessor`, so, right now, when we call `Dispose`, it calls `Shutdown`. For `BroadcastActivityProcessor`, in the dispose we have:
```csharp
protected override void Dispose(bool disposing)
        {
            if (disposing && !this.disposed)
            {
                foreach (var processor in this.processors)
                {
                    try
                    {
                        if (processor is IDisposable disposable)
                        {
                            disposable.Dispose();
                        }
                    }
                    catch (Exception e)
                    {
                        OpenTelemetrySdkEventSource.Log.SpanProcessorException("Dispose", e);
                    }
                }

                this.disposed = true;
            }

            base.Dispose(disposing);
        }
```

So, we are disposing each processor and then, disposing itself. The shutdown is:

```csharp
        public override Task ShutdownAsync(CancellationToken cancellationToken)
        {
            var tasks = new List<Task>();
            foreach (var processor in this.processors)
            {
                tasks.Add(processor.ShutdownAsync(cancellationToken));
            }

            return Task.WhenAll(tasks);
        }
```

in this shutdown, we are not checking if we already did the dispose. So, this is why we have two shutdowns for each processor that we have more then one processor added.

IF we have just one, in tracerprovidersdk, we will have just the myactivityprocessor. When we have more then one, its "transformed" in the broadcastactivityprocessor.

Reply:
I've got a fix for the `CompositeActivityProcessor` #1017.